### PR TITLE
[clang-scan-deps][cas] Initial support for caching modules + PCH using casfs

### PR DESCRIPTION
**Pull request by benlangmuir**
When a CAS and ActionCache are available to to the scanner, we can
produce build commands for modules as well as the main TU that use
the cc1 option -fcache-compile-job to enable compilation caching. This
makes use of the newly added -fmodule-file-cache-key= to support
creating commands that import a given module via an action cache key
without needing that module to have been built yet. This means that
unlike with PCH, we do not need to scan commands that use modules in
dependency order with respect to building those modules, and we do not
need to read the pcm files back from disk repeatedly in order to ingest
them into a casfs. Like any scanned explicit modules build, the modules
must still be *built* in dependency order by the build system, or else
the action cache lookup will fail in the same places that an uncached
build would fail due to a missing .pcm on disk.

Also extend the support for caching module compilations to also support
mixing PCH + modules. The PCH itself still requires scanning in
dependency order (scan PCH, build PCH, scan TU that uses PCH), but the
module dependencies of the PCH and the TU phases do not add any
additional constraints on scanning.

This PR contains the primary changes needed to build modules files
via the cache. However, there are some additional changes that will be
needed before this is practical to use:

* Support prefix mapping for paths
* Thinning out the set of files ingested into the casfs to reduce
  spurious cache misses - in particular, removing unnecessary volatile
  files such as the implicit module pcm files created by the scanner.

### CODE REVIEW

**Comment by benlangmuir** [comment]
Fixes
* Added back accidentally removed `#include "llvm/ADT/StringSet.h"`
* Fixed bug where we reused implicit pcm files built by the scanner despite changing whether caching was enabled or not

**Comment by benlangmuir** [comment]
@swift-ci please test

**Comment by benlangmuir** [comment]
All the dependencies landed. Rebased and marked ready for review.

**Comment by benlangmuir** [comment]
Review ping

**Comment by benlangmuir** [comment]
@swift-ci please test

File: clang/include/clang/Basic/Module.h [hunk 1]
```diff
@@ -189,6 +189,10 @@ class alignas(8) Module {
   /// The \c ActionCache key for this module, if any.
   Optional<std::string> ModuleCacheKey;
 
+  /// The CAS filesystem root ID for implicit modules built with the dependency
+  /// scanner, if any.
+  Optional<std::string> CASFileSystemRootID;
```
Comment (jansvoboda11) [inline clang/include/clang/Basic/Module.h:195]: I don't think this belongs to `Module`. This class tries to describe features of the module, not how it's built.

**Review by jansvoboda11 — CHANGES_REQUESTED** [review]

File: clang/include/clang/Basic/Module.h (hunk 1 above)
Comment (benlangmuir) [reply clang/include/clang/Basic/Module.h:195]: Do you have another suggestion? This is directly tied to the module and is stored in the pcm file specifically for modules.

Comment (jansvoboda11) [reply clang/include/clang/Basic/Module.h:195]: We could introduce `CASManager` and stash CAS-related state in there.

PCM files contain information that's tied to the module but end up being deserialized into other objects (`ASTContext`, `Preprocessor`, `FileManager`, `SourceManager` etc.).

Comment (benlangmuir) [reply clang/include/clang/Basic/Module.h:195]: Summary of offline discussion:
* Ben would prefer not to have a CASManager-type object, because there isn't really a single "thing" that is being managed - the CAS has different uses in different places, so this would be a grab bag of state.
* We don't like having build/serialization related state in Module itself. That includes these cas fields but also the existing ASTFile and ASTFileSignature.
* We liked the idea of a ModuleBuildManager that would extract specifically these Module fields.  Ben tried to make this refactoring, but it turns out to be fairly messy due to how these fields are used throughout the compiler -- Examples
  * libclang API that assume Module can retrieve ASTFile, which would necessitate a new fat object be allocated an managed on the CXTU, 
  * CodeGen is designed to be an ASTConsumer rather than having a lot of direct access to CompilerInstance objects, but it requires the ASTFile for debug info.
* For now, I would like to leave the cas-related fields here. If we ever do succeed in factoring out ASTFile I think that would be the right time to move these as well.

**Comment by benlangmuir** [comment]
@swift-ci please test

**Comment by benlangmuir** [comment]
Rebased.  Does anyone have additional feedback?

**Comment by benlangmuir** [comment]
@swift-ci please test

File: clang/include/clang/Frontend/CompilerInstance.h [hunk 2]
```diff
@@ -175,6 +175,12 @@ class CompilerInstance : public ModuleLoader {
   /// The list of active output files.
   std::list<llvm::vfs::OutputFile> OutputFiles;
 
+  /// An optional callback function used to wrap all FrontendActions
+  /// produced to generate imported modules before they are executed.
+  std::function<std::unique_ptr<FrontendAction>(
```
Comment (jansvoboda11) [inline clang/include/clang/Frontend/CompilerInstance.h]: Nit: Can we extract this into a typedef?

File: clang/lib/Serialization/ASTReader.cpp [hunk 3]
```diff
@@ -5196,17 +5200,25 @@ bool ASTReader::readASTFileControlBlock(
...
+    ASTReaderListener &Listener, bool ValidateDiagnosticOptions,
+    InMemoryModuleCache *ModuleCache) {
   // Open the AST file.
-  // FIXME: This allows use of the VFS; we do not allow use of the
-  // VFS when actually loading a module.
-  auto Buffer = FileMgr.getBufferForFile(Filename);
+  std::unique_ptr<llvm::MemoryBuffer> OwnedBuffer;
+  llvm::MemoryBuffer *Buffer = nullptr;
+  if (ModuleCache)
+    Buffer = ModuleCache->lookupPCM(Filename);
```
Comment (jansvoboda11) [inline clang/lib/Serialization/ASTReader.cpp]: This seems generally useful even outside of this patch, right? Could we extract this into a separate patch so it's easier to upstream?

**Review by jansvoboda11 — COMMENTED** [review]
I'm finding this hard to review. I might be able to give more constructive feedback if this gets split into more granular commits.

File: clang/lib/Serialization/ASTReader.cpp (hunk 3 above)
Comment (benlangmuir) [reply clang/lib/Serialization/ASTReader.cpp]: Good idea, split out and improved it: https://reviews.llvm.org/D138160

**Comment by benlangmuir** [comment]
* Split out two commits. One can go upstream (https://reviews.llvm.org/D138160) and I will rebase this before merging.  The other is still in this PR, but it's a standalone commit ([[clang][cas] Add a real diagnostic for missing root-id in cache key](https://github.com/apple/llvm-project/pull/5294/commits/fac2d3181e6a9c4e3c8822c7397147b6429881fe)).
* Added typedef for the module action wrapper callback

**Comment by benlangmuir** [comment]
@swift-ci please test

**Comment by benlangmuir** [comment]
@swift-ci please test

**Review by jansvoboda11 — APPROVED** [review]
LGTM.

**Comment by benlangmuir** [comment]
@swift-ci please test
